### DISCUSSION TITLE
Database design

### DISCUSSION THREAD

**Discussion by AndreyDmitriev39r** [Bill-1991/Scheduler#23] [General]
We should start thinking about how exactly all data needed will be stored and structured.

**Comment by Bill-1991** [comment]
Yeah we should...for this phase of the project we just need a table with name and free columns...database will be changing a lot down the road, as timezone and notify comes into play and much more after hours will be implemented.
We can maybe create 4 columns (name, free, timezone, notify) now and leave the last 2 as null for now.
What do you think?

**Comment by AndreyDmitriev39r** [comment]
> Yeah we should...for this phase of the project we just need a table with name and free columns...database will be changing a lot down the road, as timezone and notify comes into play and much more after hours will be implemented. We can maybe create 4 columns (name, free, timezone, notify) now and leave the last 2 as null for now. What do you think?

Columns are part of table.
Database itself is a bunch of related tables.

`The last 2` - I don't get what it is about.

`Timezone and notify comes into play` I get that you are referring to the times when it will come to coding those up.
But in some sense it it came into play already. We have feature list where it is mentioned. So it is some info database scheming can be based on.

`We can maybe create 4 columns (name, free, timezone, notify) now and leave the last 2 as null for now.`
Yes we can create some generic db. But, again it should be described and done in more detail. For example, you did not mentioned id-s. Names can be same theoretically, they can not be unique identifiers.

**Reply by Bill-1991** [reply]
Ok cool yeah i forgot ids...a 5 column table (id, name, free, timezone, notify) will do the job then for now.
Table relation will be brought up when we get to hours section where each hour of the day needs to be related to every single name in the db.

**Reply by AndreyDmitriev39r** [reply]
I still think we need more detailed database schema before jumping into database-related commands.

Otherwise we will run into such flow:

1. Suppose we have some generic database.
2. Then we implement bot with database interactions
3. Then, at some point we decide to change database.
4. Then we need refactor all the bot with database interactions
5. Then 3 and 4 are repeating several times

While some refactoring is inevitable, especially considering our experience lol, it can be minimized by thinking through how and why data storage should look like.
I would like to have ERD at least.

**Reply by Bill-1991** [reply]
![IMG_20240206_234833](https://github.com/Bill-1991/Scheduler/assets/107632771/0fe74d0c-8c13-458e-ae54-91f7ddec512d)

**Reply by Bill-1991** [reply]
Do we need to give the schema of the tables or is this just enough?

**Reply by Bill-1991** [reply]
All values can be null except for ids...Basically this must be like that...If a user inserts only his name in the database, then the user will be inserted with everything on null, dates and hours won't have any records at all for this user until he specifies them

**Comment by AndreyDmitriev39r** [comment]
> All values can be null except for ids...Basically this must be like that...If a user inserts only his name in the database, then the user will be inserted with everything on null, dates and hours won't have any records at all for this user until he specifies them

How name is the null tho? If only id is not nullable?

**Reply by Bill-1991** [reply]
And name sorry...Every other than these 2 is an option so if it is null it won't change until messed, so it comes along good with the options part
